### PR TITLE
Add csaf_validator

### LINKED ISSUE #258: Complete and add remote_validator_client
In branch https://github.com/csaf-poc/csaf_distribution/tree/remote_validator_client
there is an experimental testing tool to ask a service like https://github.com/secvisogram/csaf-validator-service .

Would need some documentation, better command line options (like --version, long options) and
displaying the diagnostic coming from the service,  to make it part of the release.

### PR DESCRIPTION
**Pull request by s-l-teichmann**
Add a `csaf_validator` tool.

Resolves #258

### CODE REVIEW

File: cmd/csaf_validator/main.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			return fmt.Errorf("remote validation of %q failed: %w",
+				file, err)
+		}
+		var passes string
+		if validate {
+			passes = "passes"
+		} else {
+			passes = "does not pass"
+		}
+		fmt.Printf("%q %s remote validation.\n", file, passes)
```
Comment (tschmidtb51) [inline cmd/csaf_validator/main.go:103]: Please add an option to see the results of the remote validator (as one could use that to run optional or informative tests).

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my comments.

File: cmd/csaf_validator/main.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	files, err := parser.Parse()
+	errCheck(err)
+
+	if opts.Version {
+		fmt.Println(util.SemVersion)
+		return
+	}
+
+	if len(files) == 0 {
+		log.Println("No domains given.")
```
[the remark this answers is not in the document]
Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go]: Resolved with 6a60e8d8ce839b4f4a58cc493569974f59bebe51

File: cmd/csaf_validator/main.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+			}
+		} else {
+			fmt.Printf("%s passes the schema validation.\n", file)
+		}
+		// Validate against remote validator.
+		validate, err := validator.Validate(doc)
+		if err != nil {
+			return fmt.Errorf("remote validation of %q failed: %w",
+				file, err)
+		}
```
[the remark this answers is not in the document]
Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go]: Resolved with 6a60e8d8ce839b4f4a58cc493569974f59bebe51

File: cmd/csaf_validator/main.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+			for _, vErr := range validationErrs {
+				fmt.Printf("  * %s\n", vErr)
+			}
+		} else {
+			fmt.Printf("%s passes the schema validation.\n", file)
+		}
+		// Validate against remote validator.
+		validate, err := validator.Validate(doc)
+		if err != nil {
+			return fmt.Errorf("remote validation of %q failed: %w",
```
[the remark this answers is not in the document]
Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go]: Resolved with 6a60e8d8ce839b4f4a58cc493569974f59bebe51

File: cmd/csaf_validator/main.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		if len(validationErrs) > 0 {
+			fmt.Printf("schema validation errors of %q\n", file)
+			for _, vErr := range validationErrs {
+				fmt.Printf("  * %s\n", vErr)
+			}
+		} else {
+			fmt.Printf("%s passes the schema validation.\n", file)
+		}
+		// Validate against remote validator.
```
[the remark this answers is not in the document]
Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go:96]: Resolved with a8493c0dd28f8995a873c42b1f372aab29e56806

File: cmd/csaf_validator/main.go (hunk 1 above)
Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go]: The [remote validation code](https://github.com/csaf-poc/csaf_distribution/blob/0c398199303a2bb0ddc76a5046933735789ef115/csaf/remotevalidation.go#L97) currently does not handle details of the validation.

To do so this code has to be extended.

Comment (tschmidtb51) [reply cmd/csaf_validator/main.go]: @s-l-teichmann Thanks for the hint. We'll track that as new issue #324. If you think that it's in the threshold - feel free to do it in this PR.

Comment (s-l-teichmann) [reply cmd/csaf_validator/main.go]: I would prefer to tackle this in another PR. 

Adding this feature would involve changing the internal caching of the remote results which leads to
to the risk to break the library. Breaking the library could endanger
the other uses in the provider, the downloader, etc.

File: cmd/csaf_validator/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply cmd/csaf_validator/main.go:96]: Only partially resolved. Tracking as new issue #325. Using this version to get it merged.

**Review by tschmidtb51 — APPROVED** [review]
LGTM and works.

**Comment by tschmidtb51** [comment]
@bernhardreiter Ready for you to go ;-)

**Review by bernhardreiter — APPROVED** [review]
The validator itself looks fine. It is not yet distributed for windows, so splitting out #326 .

**Pull request merged**